ros/angel_system_nodes/angel_system_nodes/voice_activity_detector.py

```diff
@@ -0,0 +1,368 @@
+                # audio is too long, publish anyways.
+                self._handle_publication(audio_data, 
+                                         split_timestamp,
+                                         self.bytes_per_sample, num_channels,
+                                         sample_rate)
+            else:
+                return
+
+            # Update the accumulated audio stream to remove any split data.
+            self._split_audio(split_timestamp, num_channels, sample_rate)
+            if not self.debug_mode:
+                temp_file.close()
+
+    def _split_audio(self, split_timestamp, num_channels, sample_rate):
+        '''
+        Split the accumulated audio given a timestamp for the split.
+        Also updates the accumulated timestamps
+        '''
+        split_idx = self._time_to_index(split_timestamp, sample_rate)
+        split_byte_idx = self._time_to_byte_index(split_timestamp,
+                    self.bytes_per_sample, num_channels, sample_rate)
+        with self.audio_stream_lock:
+            old_duration = self.accumulated_audio_duration
+            self.accumulated_audio = self.accumulated_audio[split_byte_idx:]
+            self.accumulated_timestamps =\
+                self.accumulated_timestamps[split_idx:]
+            self.accumulated_audio_duration -= split_timestamp
+            self.log.info(f"Audio data chopped from {old_duration} " +\
+                                    f"-> {self.accumulated_audio_duration}s...")
+    
+    def vad_server_request(self, file):
+        '''
+        Encloses the VAD server request logic.
+        '''
+        with open(file.name, 'rb') as f:
+            self.log.info("Querying Columbia VAD server...")
+            response = requests.post(self._vad_server_url,
+                                    files={'audio_data': f})
+            segments = json.loads(response.content)['segments']
+            self.log.info(f"Received VAD response: {segments}")
+            return segments
+    
+    def _handle_publication(self, audio_data, split_timestamp,
+                     sample_byte_length, num_channels, sample_rate):
+        
+        # Obtain the audio "prefix" with detected voice activity (before split).
+        audio_prefix = None
+        split_byte_idx = self._time_to_byte_index(split_timestamp,
+                sample_byte_length, num_channels, sample_rate)
+        with self.audio_stream_lock:
+            audio_prefix = audio_data[:split_byte_idx]
+            self.log.info(f"Split @ {split_timestamp}s: ({split_byte_idx})")            
+            if self.debug_mode:
+                self.log.info("Writing VOCAL ACTIVITY DETECTED prefix of " +\
+                          "currently accumulated audio data to \"split\" " +\
+                          "temporary file.")
+                self._create_temp_audio_file(
+                    audio_prefix, sample_rate, num_channels,
+                    prefix=f"split-{self.split_counter}-")
```

Comment:
@Purg @joshanderson-kw 
We evaluated this on the ros bag that @Purg most recently provided, which contained ~30s of audio. However, the VAD .wav file splits created here (when debugging mode is turned on) are of a noticeably worse quality than the source audio .wav file that accompanied this ros bag. My understanding is this source .wav file was recorded and then placed in the ros bag file. This leads me to suspect possible downsampling of audio data placed in a ROS Bag. Does this observation align with either of your understandings/expectations? We want to make sure there is no unexpected bug that is causing this behavior.

_**Remarkably & most importantly, we noticed no drop in performance on ASR.**_